src/civetweb.c

```diff
@@ -17060,6 +17090,26 @@ mg_start(const struct mg_callbacks *callbacks,
 		return NULL;
 	}
 
+	/* listen_ctrl_sd is used by mg_stop to make master thread
+	   break out safely from i/o wait in poll instead of relying on 
+	   frequent poll timeout to wake up and check the stop flag.
+	*/
+	ctx->listen_ctrl_sd = socket(AF_INET, SOCK_DGRAM, 0);
+	if(ctx->listen_ctrl_sd > 0) {
+		struct sockaddr_in saddr;
+		memset(&saddr, 0, sizeof(saddr));
+		saddr.sin_family = AF_INET;
+		saddr.sin_port = htons(0);
+		saddr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+		fcntl(ctx->listen_ctrl_sd, F_SETFD, FD_CLOEXEC);
+		set_non_blocking_mode(ctx->listen_ctrl_sd);
+		if(bind(ctx->listen_ctrl_sd, (struct sockaddr *)&saddr,
+                                                   sizeof(saddr)) < 0) {
+			close(ctx->listen_ctrl_sd);
```

Comment:
We need to handle error by returning null, and make sure other resources are properly cleared.